File: app/src/main/java/com/stolz/connect/ui/addedit/AddEditViewModel.kt
package com.stolz.connect.ui.addedit

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.stolz.connect.data.preferences.NotificationPreferences
import com.stolz.connect.data.repository.ConnectionRepository
import com.stolz.connect.domain.model.ConnectionMethod
import com.stolz.connect.domain.model.ScheduledConnection
import com.stolz.connect.util.ValidationUtils
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.util.Calendar
import java.util.Date
import javax.inject.Inject

@HiltViewModel
class AddEditViewModel @Inject constructor(
    private val connectionRepository: ConnectionRepository,
    private val notificationPreferences: NotificationPreferences,
    savedStateHandle: SavedStateHandle
) : ViewModel() {
    
    private val connectionId: Long? = savedStateHandle.get<Long>("connectionId")?.takeIf { it > 0 }
    
    private val _uiState = MutableStateFlow(
        AddEditUiState(
            contactName = "",
            contactPhoneNumber = null,
            contactEmail = null,
            contactPhotoUri = null,
            avatarColor = null,
            contactId = null,
            reminderFrequencyDays = 7,
            preferredMethod = ConnectionMethod.BOTH,
            reminderTime = null,
            notes = null,
            birthday = null,
            promptOnBirthday = true
        )
    )
    val uiState: StateFlow<AddEditUiState> = _uiState.asStateFlow()
    
    private val _saveResult = MutableStateFlow<SaveResult?>(null)
    val saveResult: StateFlow<SaveResult?> = _saveResult.asStateFlow()

    private val _duplicateCandidates = MutableStateFlow<List<com.stolz.connect.domain.model.ScheduledConnection>?>(null)
    val duplicateCandidates: StateFlow<List<com.stolz.connect.domain.model.ScheduledConnection>?> = _duplicateCandidates.asStateFlow()

    init {
        android.util.Log.d("AddEditViewModel", "Init: connectionId from savedStateHandle = ${savedStateHandle.get<Long>("connectionId")}, filtered = $connectionId")
        if (connectionId != null && connectionId > 0) {
            loadConnection(connectionId)
        } else {
            // Set default next reminder date to today and default reminder time from settings
            val calendar = Calendar.getInstance()
            _uiState.value = _uiState.value.copy(
                nextReminderDate = calendar.time,
                reminderTime = notificationPreferences.getDefaultReminderTime()
            )
        }
    }

    /** Default reminder time from settings (e.g. "10:00"). Used when connection has no reminder time set. */
    fun getDefaultReminderTime(): String = notificationPreferences.getDefaultReminderTime()
    
    private fun loadConnection(id: Long) {
        viewModelScope.launch {
            val connection = connectionRepository.getConnectionById(id)
            connection?.let {
                _uiState.value = AddEditUiState(
                    contactName = it.contactName,
                    contactPhoneNumber = it.contactPhoneNumber,
                    contactEmail = it.contactEmail,
                    contactPhotoUri = it.contactPhotoUri,
                    avatarColor = it.avatarColor,
                    contactId = it.contactId,
                    reminderFrequencyDays = it.reminderFrequencyDays,
                    preferredMethod = it.preferredMethod,
                    reminderTime = it.reminderTime,
                    notes = it.notes,
                    birthday = it.birthday,
                    promptOnBirthday = it.promptOnBirthday,
                    nextReminderDate = it.nextReminderDate
                )
                ensurePreferredMethodValid()
            }
        }
    }
    
    fun updateContactName(name: String) {
        _uiState.value = _uiState.value.copy(contactName = name)
    }
    
    fun updateContactPhoneNumber(phone: String?) {
        val cleaned = phone?.takeIf { it.isNotBlank() }
        _uiState.value = _uiState.value.copy(contactPhoneNumber = cleaned)
        ensurePreferredMethodValid()
    }
    
    fun updateContactEmail(email: String?) {
        val cleaned = email?.takeIf { it.isNotBlank() }
        _uiState.value = _uiState.value.copy(contactEmail = cleaned)
        ensurePreferredMethodValid()
    }
    
    fun updateContactPhotoUri(photoUri: String?) {
        _uiState.value = _uiState.value.copy(contactPhotoUri = photoUri)
    }
    
    fun updateAvatarColor(color: Int?) {
        _uiState.value = _uiState.value.copy(avatarColor = color)
    }
    
    fun updateContactId(contactId: String?) {
        _uiState.value = _uiState.value.copy(contactId = contactId)
    }
    
    fun updateReminderFrequencyDays(days: Int) {
        _uiState.value = _uiState.value.copy(reminderFrequencyDays = days)
    }
    
    fun updateReminderFrequencyDaysFromString(daysString: String) {
        // Allow empty string, but default to current value if invalid
        if (daysString.isBlank()) {
            // Keep current value when field is empty
            return
        }
        val days = daysString.toIntOrNull() ?: _uiState.value.reminderFrequencyDays
        _uiState.value = _uiState.value.copy(reminderFrequencyDays = days)
    }
    
    fun updatePreferredMethod(method: ConnectionMethod) {
        _uiState.value = _uiState.value.copy(preferredMethod = method)
    }

    /** If current preferred method is invalid (e.g. Email but no email), reset to a valid choice. */
    private fun ensurePreferredMethodValid() {
        val state = _uiState.value
        val hasPhone = !state.contactPhoneNumber.isNullOrBlank()
        val hasEmail = !state.contactEmail.isNullOrBlank()
        val valid = when (state.preferredMethod) {
            ConnectionMethod.CALL, ConnectionMethod.MESSAGE -> hasPhone
            ConnectionMethod.EMAIL -> hasEmail
            ConnectionMethod.BOTH -> hasPhone || hasEmail
        }
        if (!valid) {
            val newMethod = when {
                hasPhone -> ConnectionMethod.CALL
                hasEmail -> ConnectionMethod.EMAIL
                else -> ConnectionMethod.BOTH
            }
            _uiState.value = state.copy(preferredMethod = newMethod)
        }
    }
    
    fun updateReminderTime(time: String?) {
        _uiState.value = _uiState.value.copy(reminderTime = time)
    }
    
    fun updateNotes(notes: String?) {
        _uiState.value = _uiState.value.copy(notes = notes)
    }
    
    fun updateNextReminderDate(date: Date) {
        _uiState.value = _uiState.value.copy(nextReminderDate = date)
    }
    
    fun updateBirthday(date: Date?) {
        _uiState.value = _uiState.value.copy(birthday = date)
    }
    
    fun updatePromptOnBirthday(prompt: Boolean) {
        _uiState.value = _uiState.value.copy(promptOnBirthday = prompt)
    }
    
    fun saveConnection() {
        android.util.Log.d("AddEditViewModel", "saveConnection() called")
        val state = _uiState.value
        android.util.Log.d("AddEditViewModel", "State: name=${state.contactName}, phone=${state.contactPhoneNumber}, isEdit=${connectionId != null}")
        
        // Validate fields
        if (state.contactName.isBlank() || (state.contactPhoneNumber.isNullOrBlank() && state.contactEmail.isNullOrBlank())) {
            android.util.Log.w("AddEditViewModel", "Validation failed: blank fields")
            _saveResult.value = SaveResult.Error("Please provide at least a name and either a phone number or email")
            return
        }
        
        // Validate phone number if provided
        if (!state.contactPhoneNumber.isNullOrBlank() && !ValidationUtils.isValidPhone(state.contactPhoneNumber)) {
            _saveResult.value = SaveResult.Error("Please enter a valid phone number")
            return
        }
        
        // Validate email if provided
        if (!state.contactEmail.isNullOrBlank() && !ValidationUtils.isValidEmail(state.contactEmail)) {
            _saveResult.value = SaveResult.Error("Please enter a valid email address")
            return
        }
        
        // Validate preferred method matches available data
        val hasPhone = !state.contactPhoneNumber.isNullOrBlank()
        val hasEmail = !state.contactEmail.isNullOrBlank()
        val preferredMethodValid = when (state.preferredMethod) {
            ConnectionMethod.CALL, ConnectionMethod.MESSAGE -> {
                if (!hasPhone) {
                    _saveResult.value = SaveResult.Error("Please enter a phone number to use Call or Message as preferred method")
                    return
                }
                true
            }
            ConnectionMethod.EMAIL -> {
                if (!hasEmail) {
                    _saveResult.value = SaveResult.Error("Please enter an email address to use Email as preferred method")
                    return
                }
                true
            }
            ConnectionMethod.BOTH -> {
                if (!hasPhone && !hasEmail) {
                    _saveResult.value = SaveResult.Error("Please enter at least a phone number or email address")
                    return
                }
                true
            }
        }
        
        viewModelScope.launch {
            try {
                android.util.Log.d("AddEditViewModel", "Starting save process...")
                val nextDate = state.nextReminderDate ?: Date()
                val connection = ScheduledConnection(
                    id = connectionId ?: 0,
                    contactName = state.contactName,
                    contactPhoneNumber = state.contactPhoneNumber,
                    contactEmail = state.contactEmail,
                    contactPhotoUri = state.contactPhotoUri,
                    avatarColor = state.avatarColor,
                    contactId = state.contactId,
                    reminderFrequencyDays = state.reminderFrequencyDays,
                    preferredMethod = state.preferredMethod,
                    reminderTime = state.reminderTime,
                    notes = state.notes,
                    birthday = state.birthday,
                    promptOnBirthday = state.promptOnBirthday,
                    nextReminderDate = nextDate,
                    isActive = true
                )

                if (connectionId != null) {
                    val duplicates = connectionRepository.findPotentialDuplicates(
                        state.contactName,
                        state.contactPhoneNumber,
                        state.contactEmail,
                        excludeId = connectionId
                    )
                    if (duplicates.isNotEmpty()) {
                        _duplicateCandidates.value = duplicates
                        return@launch
                    }
                    connectionRepository.updateConnection(connection)
                    kotlinx.coroutines.delay(200)
                    _saveResult.value = SaveResult.Success
                } else {
                    val duplicates = connectionRepository.findPotentialDuplicates(
                        state.contactName,
                        state.contactPhoneNumber,
                        state.contactEmail
                    )
                    if (duplicates.isNotEmpty()) {
                        _duplicateCandidates.value = duplicates
                        return@launch
                    }
                    val insertedId = connectionRepository.insertConnection(connection)
                    if (insertedId <= 0) throw Exception("Failed to insert connection")
                    kotlinx.coroutines.delay(200)
                    _saveResult.value = SaveResult.Success
                }
            } catch (e: Exception) {
                android.util.Log.e("AddEditViewModel", "Save failed with exception", e)
                _saveResult.value = SaveResult.Error(e.message ?: "Failed to save connection")
            }
        }
    }

    /** Performs insert without duplicate check (user chose "Add anyway"). */
    fun saveConnectionIgnoringDuplicates() {
        _duplicateCandidates.value = null
        val state = _uiState.value
        if (state.contactName.isBlank() || (state.contactPhoneNumber.isNullOrBlank() && state.contactEmail.isNullOrBlank())) return
        viewModelScope.launch {
            try {
                val nextDate = state.nextReminderDate ?: Date()
                val connection = ScheduledConnection(
                    id = 0,
                    contactName = state.contactName,
                    contactPhoneNumber = state.contactPhoneNumber,
                    contactEmail = state.contactEmail,
                    contactPhotoUri = state.contactPhotoUri,
                    avatarColor = state.avatarColor,
                    contactId = state.contactId,
                    reminderFrequencyDays = state.reminderFrequencyDays,
                    preferredMethod = state.preferredMethod,
                    reminderTime = state.reminderTime,
                    notes = state.notes,
                    birthday = state.birthday,
                    promptOnBirthday = state.promptOnBirthday,
                    nextReminderDate = nextDate,
                    isActive = true
                )
                val insertedId = connectionRepository.insertConnection(connection)
                if (insertedId <= 0) throw Exception("Failed to insert connection")
                kotlinx.coroutines.delay(200)
                _saveResult.value = SaveResult.Success
            } catch (e: Exception) {
                _saveResult.value = SaveResult.Error(e.message ?: "Failed to save connection")
            }
        }
    }

    /** Performs update without duplicate check (user chose "Save anyway" when editing). */
    fun updateConnectionIgnoringDuplicates() {
        _duplicateCandidates.value = null
        val state = _uiState.value
        val id = connectionId ?: return
        viewModelScope.launch {
            try {
                val nextDate = state.nextReminderDate ?: Date()
                val connection = ScheduledConnection(
                    id = id,
                    contactName = state.contactName,
                    contactPhoneNumber = state.contactPhoneNumber,
                    contactEmail = state.contactEmail,
                    contactPhotoUri = state.contactPhotoUri,
                    avatarColor = state.avatarColor,
                    contactId = state.contactId,
                    reminderFrequencyDays = state.reminderFrequencyDays,
                    preferredMethod = state.preferredMethod,
                    reminderTime = state.reminderTime,
                    notes = state.notes,
                    birthday = state.birthday,
                    promptOnBirthday = state.promptOnBirthday,
                    nextReminderDate = nextDate,
                    isActive = true
                )
                connectionRepository.updateConnection(connection)
                kotlinx.coroutines.delay(200)
                _saveResult.value = SaveResult.Success
            } catch (e: Exception) {
                _saveResult.value = SaveResult.Error(e.message ?: "Failed to save connection")
            }
        }
    }

    fun clearDuplicateCandidates() {
        _duplicateCandidates.value = null
    }

    fun clearSaveResult() {
        _saveResult.value = null
    }
}

data class AddEditUiState(
    val contactName: String,
    val contactPhoneNumber: String? = null,
    val contactEmail: String? = null,
    val contactPhotoUri: String? = null,
    val avatarColor: Int? = null,
    val contactId: String? = null,
    val reminderFrequencyDays: Int,
    val preferredMethod: ConnectionMethod,
    val reminderTime: String?,
    val notes: String?,
    val birthday: Date? = null,
    val promptOnBirthday: Boolean = true,
    val nextReminderDate: Date? = null
)

sealed class SaveResult {
    object Success : SaveResult()
    data class Error(val message: String) : SaveResult()
}

File: ios/Connect/ViewModels/AddEditViewModel.swift
import Foundation
import Combine

final class AddEditViewModel: ObservableObject {
    private let store = ConnectionStore.shared
    private let prefs = Preferences.shared

    let connectionId: Int64?
    @Published var contactName = ""
    @Published var contactPhoneNumber: String?
    @Published var contactEmail: String?
    @Published var reminderFrequencyDays = 7
    /// When user selects "Custom" frequency, this holds the day count (1–365).
    @Published var customFrequencyDays = 7
    @Published var preferredMethod: ConnectionMethod = .both
    @Published var reminderTime: String?
    @Published var notes: String?
    @Published var birthday: Date?
    @Published var promptOnBirthday = true
    @Published var nextReminderDate = Date()
    @Published var contactPhotoUri: String?
    @Published var avatarColor: Int?
    @Published var contactId: String?
    @Published var saveResult: SaveResult?
    @Published var duplicateCandidates: [ScheduledConnection]?
    /// Set after inserting a new connection so the UI can navigate to it.
    private(set) var lastInsertedId: Int64?

    /// Human-readable description of which fields matched (e.g. "Name, Phone") when duplicateCandidates is set.
    var duplicateMatchDescription: String? {
        guard let dupes = duplicateCandidates, let first = dupes.first else { return nil }
        let name = contactName.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let phoneNorm = contactPhoneNumber?.replacingOccurrences(of: "[^0-9]", with: "", options: .regularExpression)
        let emailNorm = contactEmail?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        var parts: [String] = []
        if !name.isEmpty && first.contactName.trimmingCharacters(in: .whitespaces).lowercased() == name { parts.append("Name") }
        if let p = phoneNorm, !p.isEmpty, first.contactPhoneNumber?.replacingOccurrences(of: "[^0-9]", with: "", options: .regularExpression) == p { parts.append("Phone") }
        if let e = emailNorm, !e.isEmpty, first.contactEmail?.trimmingCharacters(in: .whitespaces).lowercased() == e { parts.append("Email") }
        return parts.isEmpty ? nil : parts.joined(separator: ", ")
    }

    init(connectionId: Int64?) {
        self.connectionId = connectionId
        if let id = connectionId, id > 0, let c = store.connection(byId: id) {
            contactName = c.contactName
            contactPhoneNumber = c.contactPhoneNumber
            contactEmail = c.contactEmail
            let days = c.reminderFrequencyDays
            reminderFrequencyDays = [1, 7, 14, 30].contains(days) ? days : 0
            customFrequencyDays = [1, 7, 14, 30].contains(days) ? 7 : days
            preferredMethod = c.preferredMethod
            reminderTime = c.reminderTime
            notes = c.notes
            birthday = c.birthday
            promptOnBirthday = c.promptOnBirthday
            nextReminderDate = c.nextReminderDate
            contactPhotoUri = c.contactPhotoUri
            avatarColor = c.avatarColor
            contactId = c.contactId
        } else {
            reminderTime = prefs.defaultReminderTime
            avatarColor = AvatarColors.argbValues[Int.random(in: 0..<AvatarColors.argbValues.count)]
        }
    }

    func updateContactPhotoUri(_ path: String?) { contactPhotoUri = path }
    func updateAvatarColor(_ int: Int?) { avatarColor = int }

    var defaultReminderTime: String { prefs.defaultReminderTime }

    /// Whether preferred method is valid given current phone/email (call/message/faceTime need phone, email needs email, both needs at least one).
    func isPreferredMethodValid(hasPhone: Bool, hasEmail: Bool) -> Bool {
        switch preferredMethod {
        case .call, .message, .faceTime: return hasPhone
        case .email: return hasEmail
        case .both: return hasPhone || hasEmail
        }
    }

    /// If current preferred method is invalid (e.g. Email selected but no email), reset to a valid choice.
    func ensurePreferredMethodValid() {
        let hasPhone = (contactPhoneNumber?.trimmingCharacters(in: .whitespaces)).map { !$0.isEmpty } ?? false
        let hasEmail = (contactEmail?.trimmingCharacters(in: .whitespaces)).map { !$0.isEmpty } ?? false
        guard !isPreferredMethodValid(hasPhone: hasPhone, hasEmail: hasEmail) else { return }
        if hasPhone { preferredMethod = .call }
        else if hasEmail { preferredMethod = .email }
        else { preferredMethod = .both }
    }

    func save() {
        let name = contactName.trimmingCharacters(in: .whitespacesAndNewlines)
        let phone = contactPhoneNumber?.trimmingCharacters(in: .whitespaces).nilIfEmpty
        let email = contactEmail?.trimmingCharacters(in: .whitespaces).nilIfEmpty
        if name.isEmpty || (phone == nil && email == nil) {
            saveResult = .error("Please provide at least a name and either a phone number or email")
            return
        }
        if let p = phone, !ValidationUtils.isValidPhone(p) {
            saveResult = .error("Please enter a valid phone number")
            return
        }
        if let e = email, !ValidationUtils.isValidEmail(e) {
            saveResult = .error("Please enter a valid email address")
            return
        }
        switch preferredMethod {
        case .call, .message, .faceTime:
            if phone == nil { saveResult = .error("Please enter a phone number to use Call, Message, or FaceTime"); return }
        case .email:
            if email == nil { saveResult = .error("Please enter an email to use Email"); return }
        case .both:
            if phone == nil && email == nil { saveResult = .error("Please enter at least a phone or email"); return }
        }
        let duplicates = store.findPotentialDuplicates(name: name, phone: phone, email: email, excludeId: connectionId)
        if !duplicates.isEmpty {
            duplicateCandidates = duplicates
            return
        }
        performSave(name: name, phone: phone, email: email)
    }

    func saveAnyway() {
        duplicateCandidates = nil
        let name = contactName.trimmingCharacters(in: .whitespacesAndNewlines)
        let phone = contactPhoneNumber?.trimmingCharacters(in: .whitespaces).nilIfEmpty
        let email = contactEmail?.trimmingCharacters(in: .whitespaces).nilIfEmpty
        performSave(name: name, phone: phone, email: email)
    }

    func clearDuplicateCandidates() { duplicateCandidates = nil }
    func clearSaveResult() { saveResult = nil }

    /// Fill form from a contact picked from the device (name, phone, email).
    func applyPickedContact(name: String, phone: String?, email: String?) {
        contactName = name
        contactPhoneNumber = phone
        contactEmail = email
    }

    private func performSave(name: String, phone: String?, email: String?) {
        let days = reminderFrequencyDays == 0 ? min(365, max(1, customFrequencyDays)) : reminderFrequencyDays
        var conn = ScheduledConnection(
            id: connectionId ?? 0,
            contactName: name,
            contactPhoneNumber: phone,
            contactEmail: email,
            contactPhotoUri: contactPhotoUri,
            avatarColor: avatarColor,
            contactId: contactId,
            reminderFrequencyDays: days,
            preferredMethod: preferredMethod,
            reminderTime: reminderTime,
            nextReminderDate: nextReminderDate,
            notes: notes?.nilIfEmpty,
            birthday: birthday,
            promptOnBirthday: promptOnBirthday,
            isActive: true
        )
        if let id = connectionId, id > 0 {
            conn.id = id
            store.update(conn)
            if prefs.notificationsEnabled {
                NotificationService.schedule(connection: conn, defaultReminderTime: prefs.defaultReminderTime)
            }
        } else {
            let insertedId = store.insert(conn)
            lastInsertedId = insertedId
            if let photoPath = contactPhotoUri, !photoPath.isEmpty {
                PhotoStorage.replacePhotoAfterInsert(oldPath: photoPath, newConnectionId: insertedId)
                let newPath = "connection_photos/\(insertedId).jpg"
                if var updated = store.connection(byId: insertedId) {
                    updated.contactPhotoUri = newPath
                    store.update(updated)
                }
            }
            if prefs.notificationsEnabled, let c = store.connection(byId: insertedId) {
                NotificationService.schedule(connection: c, defaultReminderTime: prefs.defaultReminderTime, showIfDueNow: false)
            }
        }
        saveResult = .success
    }
}

enum SaveResult: Equatable {
    case success
    case error(String)
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}
